### PR TITLE
dvc: optmize PathInfo.isin()

### PR DESCRIPTION
**Pull request by Suor**
Should help with #2203

### CODE REVIEW

File: dvc/path_info.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 
 class PathInfo(pathlib.PurePath):
+    __slots__ = ()
```
Comment (efiop) [inline dvc/path_info.py]: Could you elaborate on what is going on here and down below, please?

Comment (Suor) [reply dvc/path_info.py]: `pathlib.PurePath` uses [slots](https://docs.python.org/3/reference/datamodel.html#slots) to make itself faster, these prevent a creation of `__dict__`, make objects smaller and speed up attribute access. When you subclass such object you need to specify `__slots__` again to force this optimization on a new class. Empty `__slots__` mean we don't add new attributes. 

`_cparts` is a list of casefolded parts of a path. `pathlib` uses this property to compare `Path` objects (`__eq__`, `__gt__`, `__lt__` and friends). We use them to compare start of `self` with `other`.

**Comment by Suor** [comment]
There are two places where `.isin()` calls come in pairs:
```python
in_o_dir = out.path_info.isin(o.path_info)
in_out_dir = o.path_info.isin(out.path_info)
if in_o_dir or in_out_dir:
    # ...
```
These might be optimized by adding PathInfo/UrlInfo.commonprefix:
```python
prefix = out.path_info.commonprefix(o.path_info)
if prefix in (out.path_info, o.path_info):
    # ...
```

Or even better replace `outs` list with some tree structure, this is the only way to get rid of quadratic complexity. But if current version is fast enough we may stop here.

**Comment by Suor** [comment]
@efiop this fails `test_get_repo_dir` with `FileNotFoundError: [WinError 206] The filename or extension is too long` seems unrelated.

File: dvc/path_info.py (hunk 1 above)
Comment (efiop) [reply dvc/path_info.py]: Thank you! Could we add these as a comment? Would be a shame to lose it in PR.

File: dvc/repo/__init__.py [hunk 2]
```diff
@@ -308,11 +308,11 @@ def graph(self, stages=None, from_directory=None):
                 for o in outs:
                     if o.path_info == out.path_info:
                         existing.append(o.stage)
-
-                    in_o_dir = out.path_info.isin(o.path_info)
-                    in_out_dir = o.path_info.isin(out.path_info)
-                    if in_o_dir or in_out_dir:
-                        raise OverlappingOutputPathsError(o, out)
+                    else:
```
Comment (efiop) [inline dvc/repo/__init__.py]: A bonus fix? :slightly_smiling_face:

Comment (efiop) [reply dvc/repo/__init__.py]: Could you explain what went wrong here, please? So `isin` returns true if path_infos are matching, and that caused the issue? Was it introduced by this patch, or did we have this bug earlier?

File: dvc/path_info.py (hunk 1 above)
Comment (Suor) [reply dvc/path_info.py]: Slots is built-in python mechanic, do you really want me to comment on it?
And there is a comment about `_cparts` below.

File: dvc/repo/__init__.py (hunk 2 above)
Comment (Suor) [reply dvc/repo/__init__.py]: Hm, looks like I changed `.isin()` semantics from strict to non-strict and that caused an error. This change is still desirable for optimization purposes.

File: dvc/path_info.py (hunk 1 above)
Comment (efiop) [reply dvc/path_info.py]: This is not at all obvious, so let's add a comment about it, so anyone who is not Python-fu master could understand it too, or at least have some hints :slightly_smiling_face:

File: dvc/repo/__init__.py (hunk 2 above)
Comment (shcheklein) [reply dvc/repo/__init__.py]: @Suor can you give a little bit more context please, what does strict vs non-strict mean?

Comment (Suor) [reply dvc/repo/__init__.py]: @shcheklein like < vs <=. It used to be `pi.isin(pi) is False` and I made it true here.

Need to revert and rename `.isin()` to make it obvious that it's strict but I can't come up with a good name.

**Pull request merged**